File: #0092.reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {

        ListNode *result = new ListNode(-1);
        result->next = head;
        ListNode *pre = result;
        ListNode *cur = head;
        int i = 1;
        for(; i < left; ++i){
            pre = pre->next;
            cur = cur->next;
        }

        ListNode *precur = pre;
        ListNode *cursave = cur;
        ListNode* curnext;
        for(; i <= right; ++i){
            curnext = cur->next;
            cur->next = pre;
            pre = cur;
            cur = curnext;
        }

        precur->next = pre;
        cursave->next = cur;
        return result->next;
    }
};

File: #0092.reverse-linked-list-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function(head, left, right) {
    let temp = new ListNode(0, head);
    let prev = temp;
    for(let i = 1; i < left; i++) {
        prev = prev.next;
    }

    let curr = prev.next;
    for (let i = left; i < right; i++) {
        let nextNode = curr.next;
        curr.next = nextNode.next;
        nextNode.next = prev.next;
        prev.next = nextNode;
    }
    return temp.next;
};
